File: os_scripts/cachyos.py
import logging
import os
import re
import shlex
import shutil
import stat
import subprocess
import sys
from pathlib import Path
from typing import TYPE_CHECKING

from rich.prompt import Confirm

if TYPE_CHECKING:
    from bankai.tui import TUI

# --- Logger and Global Setup ---
# log is now passed in from bankai.py
CURRENT_DIR = Path.cwd()

# --- Constants ---
STEPS = [
    "Base Dependencies",
    "AUR Helper (paru)",
    "Development Tools",
    "System Packages (pacman)",
    "Nerd Fonts",
    "Gaming Packages",
    "Flatpak Packages",
    "Fish Shell",
    "Git Configuration",
    "Terminal Setup",
    "KDE Tweaks",
    "MPV (uosc)",
    "System Services",
    "Fastfetch",
]


# URLs
RUSTUP_URL = "https://sh.rustup.rs"
PARU_AUR_URL = "https://aur.archlinux.org/paru.git"
UV_INSTALL_URL = "https://astral.sh/uv/install.sh"
FOUNDRY_INSTALL_URL = "https://foundry.paradigm.xyz"
UOSC_INSTALL_URL = (
    "https://raw.githubusercontent.com/tomasklaen/uosc/HEAD/installers/unix.sh"
)

# Paths
HOME = Path.home()
CARGO_HOME = HOME / ".cargo"
PARU_BUILD_DIR = Path("/tmp/paru")
STARSHIP_CONFIG_PATH = HOME / ".config" / "starship.toml"
FISH_CONFIG_DIR = HOME / ".config" / "fish"
PYENV_ROOT = HOME / ".pyenv"
GHOSTTY_CONFIG_DIR = HOME / ".config" / "ghostty"
FASTFETCH_CONFIG_DIR = HOME / ".config" / "fastfetch"

# Relative paths from project root
COMMON_DIR = CURRENT_DIR / "common"
CONFIGS_DIR = CURRENT_DIR / "configs"
HELPERS_DIR = CURRENT_DIR / "helpers"

PARU_APPLIST_PATH = COMMON_DIR / "paru_applist.txt"
FLATPAK_APPLIST_PATH = COMMON_DIR / "flatpacks_arch.txt"
CUSTOM_FISH_CONFIG_PATH = CONFIGS_DIR / "fish" / "config.fish"
CUSTOM_GHOSTTY_CONFIG_PATH = CONFIGS_DIR / "ghostty" / "config"
CUSTOM_FASTFETCH_CONFIG_PATH = CONFIGS_DIR / "fastfetch" / "config.jsonc"


# --- Helper Functions ---
def strip_ansi(text):
    """Removes ANSI escape sequences from a string."""
    ansi_escape = re.compile(r"(\x9B|\x1B\[)[0-?]*[ -/]*[@-~]")
    return ansi_escape.sub("", text)


def refresh_sudo(tui: "TUI"):
    """Refreshes the sudo timestamp, prompting for a password if needed."""
    tui.log_output("Checking sudo access. You may be prompted for your password.")
    try:
        # We need to run this interactively, so we pause the TUI
        tui.live.stop()
        subprocess.run(["sudo", "-v"], check=True)
        tui.live.start(refresh=True)
        tui.log_output("Sudo access confirmed.")
        return True
    except (subprocess.CalledProcessError, FileNotFoundError):
        tui.live.start(refresh=True)
        tui.log_output("[bold red]Failed to acquire sudo privileges.[/bold red]")
        return False


def run_command(command, tui: "TUI", cwd=None, check=True):
    """Runs a shell command and streams its output to the TUI."""
    tui.log_output(f"[dim]Executing: {command}[/dim]")

    use_shell = "|" in command or "source" in command or "&&" in command

    try:
        process = subprocess.Popen(
            command if use_shell else shlex.split(command),
            stdout=subprocess.PIPE,
            stderr=subprocess.STDOUT,
            text=True,
            cwd=cwd,
            shell=use_shell,
            bufsize=1,
            errors="replace",
        )

        with process.stdout:
            for line in iter(process.stdout.readline, ""):
                tui.log_output(strip_ansi(line.strip()))

        returncode = process.wait()

        if check and returncode != 0:
            tui.log_output(
                f"[yellow]Warning: Command '{command}' exited with code {returncode}[/yellow]"
            )
            return False
        return True

    except FileNotFoundError:
        tui.log_output(
            f"[bold red]Error: Command not found for: '{command}'[/bold red]"
        )
        return False
    except Exception as e:
        tui.log_output(f"[bold red]An unexpected error occurred: {e}[/bold red]")
        return False


def run_interactive_command(command, tui: "TUI", cwd=None):
    """Runs an interactive shell command by pausing the TUI."""
    tui.log_output(f"[dim]Running interactive command: {command}[/dim]")
    tui.live.stop()
    try:
        result = subprocess.run(command, shell=True, check=False, cwd=cwd)
        tui.live.start(refresh=True)
        if result.returncode != 0:
            tui.log_output(
                f"[yellow]Warning: Command '{command}' exited with code {result.returncode}[/yellow]"
            )
            return False
        return True
    except Exception as e:
        tui.live.start(refresh=True)
        tui.log_output(
            f"[bold red]An error occurred during interactive command: {e}[/bold red]"
        )
        return False


def command_exists(command):
    """Checks if a command exists in the system's PATH."""
    return shutil.which(command) is not None


def prompt_user(tui: "TUI", prompt, default=False):
    """Asks the user a yes/no question within the TUI context."""
    tui.live.stop()
    try:
        response = Confirm.ask(prompt, default=default)
    except KeyboardInterrupt:
        tui.live.start(refresh=True)
        tui.log_output("\n[yellow]Operation cancelled by user.[/yellow]")
        raise
    tui.live.start(refresh=True)
    return response


# --- Installation Functions ---
def install_base_dependencies(tui: "TUI"):
    """Installs base-devel and Rust."""
    tui.log_output("Updating system and installing base-devel...")
    run_command("sudo pacman -Syu base-devel --noconfirm", tui)

    tui.log_output("Installing Rust via rustup...")
    run_command(f"curl {RUSTUP_URL} -sSf | sh -s -- -y", tui)
    os.environ["PATH"] += f":{CARGO_HOME}/bin"

    if command_exists("pyenv"):
        tui.log_output("Configuring Pyenv for Fish...")
        run_command(
            f'fish -c "set -Ux PYENV_ROOT {PYENV_ROOT}; fish_add_path {PYENV_ROOT}/bin"',
            tui,
        )


def install_aur_helper(tui: "TUI"):
    """Installs Paru AUR helper if not present."""
    if command_exists("paru"):
        tui.log_output("Paru is already installed.")
        return
    tui.log_output("Installing paru...")
    run_interactive_command(
        (
            f"git clone {PARU_AUR_URL} {PARU_BUILD_DIR} && "
            f"cd {PARU_BUILD_DIR} && "
            "makepkg -si"
        ),
        tui,
    )


def install_dev_tools(tui: "TUI"):
    """Installs development tools like uv and Foundry."""
    if not command_exists("uv"):
        tui.log_output("Installing uv...")
        run_command(f"curl -LsSf {UV_INSTALL_URL} | sh", tui)
    else:
        tui.log_output("uv already installed.")

    if not command_exists("foundryup"):
        tui.log_output("Installing Foundry...")
        run_command(f"curl -L {FOUNDRY_INSTALL_URL} | bash", tui)
    else:
        tui.log_output("Foundry (foundryup) already installed.")


def install_packages_from_file(tui: "TUI", file_path, installer_cmd):
    """Installs packages from a text file using a specified command."""
    packages_file = Path(file_path)
    if not packages_file.is_file():
        tui.log_output(
            f"[yellow]Warning: Package file not found at {packages_file}[/yellow]"
        )
        return

    packages = [
        line.strip()
        for line in packages_file.read_text().splitlines()
        if line.strip() and not line.startswith("#")
    ]
    if packages:
        tui.log_output(f"Installing {len(packages)} packages...")
        # Use --batchinstall for non-interactive scripting
        run_command(f"{installer_cmd} --batchinstall {' '.join(packages)}", tui)


# --- Configuration Functions ---
def configure_fish_shell(tui: "TUI"):
    """Installs and configures the Fish shell and plugins."""
    if not command_exists("fish"):
        tui.log_output("Installing Fish shell...")
        run_command("paru -S fish --noconfirm", tui)

    if prompt_user(tui, "Set Fish as the default shell?", default=True):
        fish_path = shutil.which("fish")
        if fish_path:
            tui.log_output("Setting Fish as the default shell...")
            # This is interactive, so we need to handle it specially
            run_interactive_command(f"chsh -s {fish_path}", tui)
        else:
            tui.log_output("[yellow]Warning: Could not find fish executable.[/yellow]")

    tui.log_output("Installing Fisher and plugins...")
    fisher_plugins = [
        "jorgebucaran/fisher",
        "meaningful-ooo/sponge",
        "jorgebucaran/nvm.fish",
        "franciscolourenco/done",
        "joseluisq/gitnow@2.12.0",
    ]
    for plugin in fisher_plugins:
        run_command(f'fish -c "fisher install {plugin}"', tui, check=False)

    if command_exists("starship"):
        tui.log_output("Configuring Starship prompt...")
        run_command(
            f"starship preset nerd-font-symbols -o {STARSHIP_CONFIG_PATH}",
            tui,
            check=False,
        )

    FISH_CONFIG_DIR.mkdir(parents=True, exist_ok=True)
    if CUSTOM_FISH_CONFIG_PATH.is_file():
        shutil.copy(CUSTOM_FISH_CONFIG_PATH, FISH_CONFIG_DIR / "config.fish")
        tui.log_output("Copied custom fish config.")


def configure_terminals(tui: "TUI"):
    """Configures Kitty, Ghostty, and Alacritty."""
    for script_name in ["kitty.sh", "alacritty.sh"]:
        script_path = HELPERS_DIR / script_name
        if script_path.is_file():
            tui.log_output(f"Running {script_name}...")
            st = script_path.stat()
            script_path.chmod(st.st_mode | stat.S_IEXEC)
            run_command(f"bash {script_path} {str(CURRENT_DIR)}", tui)
        else:
            tui.log_output(f"Helper script not found: {script_path}")

    tui.log_output("Configuring Ghostty terminal...")
    GHOSTTY_CONFIG_DIR.mkdir(parents=True, exist_ok=True)
    if CUSTOM_GHOSTTY_CONFIG_PATH.is_file():
        shutil.copy(CUSTOM_GHOSTTY_CONFIG_PATH, GHOSTTY_CONFIG_DIR / "config")
        tui.log_output("Copied custom Ghostty config.")


def enable_services(tui: "TUI"):
    """Prompts to enable and configure system services like Bluetooth and Docker."""
    if prompt_user(tui, "Enable Bluetooth?", default=False):
        tui.log_output("Enabling Bluetooth service...")
        run_command("sudo systemctl enable --now bluetooth", tui, check=False)

    if command_exists("docker") and prompt_user(tui, "Enable Docker?", default=True):
        tui.log_output("Enabling and starting Docker service...")
        run_command("sudo systemctl enable --now docker", tui, check=False)
        user = os.getlogin()
        run_command(f"sudo usermod -aG docker {user}", tui, check=False)
        tui.log_output(
            f"[yellow]User {user} added to docker group. Please log out and back in.[/yellow]"
        )


def configure_fastfetch(tui: "TUI"):
    """Configures Fastfetch by copying the configuration file."""
    if not command_exists("fastfetch"):
        tui.log_output(
            "[yellow]Warning: Fastfetch command not found. Skipping configuration.[/yellow]"
        )
        return

    tui.log_output("Configuring Fastfetch...")
    FASTFETCH_CONFIG_DIR.mkdir(parents=True, exist_ok=True)
    if CUSTOM_FASTFETCH_CONFIG_PATH.is_file():
        shutil.copy(CUSTOM_FASTFETCH_CONFIG_PATH, FASTFETCH_CONFIG_DIR / "config.jsonc")
        tui.log_output("Fastfetch user config updated.")
    else:
        tui.log_output(
            f"[yellow]Warning: Custom Fastfetch config not found at {CUSTOM_FASTFETCH_CONFIG_PATH}[/yellow]"
        )


def configure_kde(tui: "TUI"):
    """Applies KDE-specific tweaks and configurations."""
    tui.log_output("Applying KDE Connect fix...")
    run_command("sudo iptables -I INPUT -p tcp --dport 1714:1764 -j ACCEPT", tui)
    run_command("sudo iptables -I INPUT -p udp --dport 1714:1764 -j ACCEPT", tui)
    if command_exists("ufw"):
        run_command("sudo ufw allow 1714:1764/udp", tui)
        run_command("sudo ufw allow 1714:1764/tcp", tui)
        run_command("sudo ufw reload", tui)

    if prompt_user(
        tui, "Install KDE Force Blur effect (requires build)?", default=False
    ):
        tui.log_output("Installing prerequisites for KDE Force Blur...")
        run_command(
            "paru -S base-devel git extra-cmake-modules qt6-tools --noconfirm", tui
        )
        tui.log_output("Cloning and building KDE Force Blur...")
        run_command(
            "cd /tmp && "
            "git clone https://github.com/taj-ny/kwin-effects-forceblur && "
            "cd kwin-effects-forceblur && "
            "mkdir build && cd build && "
            "cmake ../ -DCMAKE_INSTALL_PREFIX=/usr && "
            "make && sudo make install",
            tui,
            check=False,
        )


# --- Main Execution ---
def run_setup(tui: "TUI", log: "logging.Logger"):
    """Main execution flow for the CachyOS setup script."""
    with tui.task("Installing base dependencies..."):
        install_base_dependencies(tui)

    with tui.task("Installing AUR helper..."):
        install_aur_helper(tui)

    with tui.task("Installing development tools..."):
        install_dev_tools(tui)

    tui.log_output("Preparing for package installation...")
    if not refresh_sudo(tui):
        tui.log_output(
            "[bold red]Sudo authentication failed. Skipping sudo-dependent installations.[/bold red]"
        )
    else:
        with tui.task("Installing system packages..."):
            # This is now handled interactively to prevent getting stuck
            packages_file = Path(PARU_APPLIST_PATH)
            if not packages_file.is_file():
                tui.log_output(
                    f"[yellow]Warning: Package file not found at {packages_file}[/yellow]"
                )
            else:
                packages = [
                    line.strip()
                    for line in packages_file.read_text().splitlines()
                    if line.strip() and not line.startswith("#")
                ]
                if packages:
                    tui.log_output(
                        f"Handing over to paru to install {len(packages)} packages..."
                    )
                    command = f"paru -S --sudoloop {' '.join(packages)}"
                    run_interactive_command(command, tui)

        with tui.task("Installing Nerd Fonts..."):
            run_command(
                "sudo pacman -S $(pacman -Sgq nerd-fonts) --noconfirm", tui, check=False
            )

        if prompt_user(tui, "Enable gaming configuration?", default=False):
            with tui.task("Installing gaming packages..."):
                run_command(
                    "paru -S cachyos-gaming-meta cachyos-gaming-applications protonup-rs-bin --noconfirm",
                    tui,
                )
        else:
            # Advance progress bar even if skipped
            tui.skip_task("Install gaming packages")

    with tui.task("Installing Flatpak packages..."):
        # These don't require sudo, so they can run regardless.
        install_packages_from_file(
            tui,
            FLATPAK_APPLIST_PATH,
            "flatpak install --user -y flathub",
        )

    with tui.task("Configuring Fish shell..."):
        configure_fish_shell(tui)

    if prompt_user(tui, "Configure git?", default=True):
        with tui.task("Configuring git..."):
            script_path = HELPERS_DIR / "configure_git.py"
            if script_path.is_file():
                run_interactive_command(f"python3 {script_path}", tui)
            else:
                tui.log_output(
                    f"[yellow]Warning: Helper script not found: {script_path}[/yellow]"
                )
    else:
        tui.skip_task("Configure git")

    with tui.task("Configuring terminals..."):
        configure_terminals(tui)

    with tui.task("Configuring KDE..."):
        configure_kde(tui)

    with tui.task("Installing uosc for MPV..."):
        run_command(f'bash -c "$(curl -fsSL {UOSC_INSTALL_URL})"', tui)

    with tui.task("Enabling system services..."):
        enable_services(tui)

    with tui.task("Configuring Fastfetch..."):
        configure_fastfetch(tui)

    tui.log_output(
        "[bold green]CachyOS setup finished. Please restart your terminal or log out.[/bold green]"
    )

File: src/os_scripts/cachyos.js
import fs from "node:fs";
import { homedir } from "node:os";
import path from "node:path";
import { withSpinner } from "../common/ui.js";
import {
	commandExists,
	log,
	promptUser,
	runCommand,
	safeCopyFile,
} from "../common/utils.js";
import { configureAgentOs } from "../helpers/configure_agent_os.js";
import { configureAudio } from "../helpers/configure_audio.js";
import { configureBash } from "../helpers/configure_bash.js";
import { configureClaude } from "../helpers/configure_claude.js";
import { configureClaudeStayAwake } from "../helpers/configure_claude_stay_awake.js";
import { configureCodex } from "../helpers/configure_codex.js";
import { configureMimeapps } from "../helpers/configure_mimeapps.js";
import { configurePrWatch } from "../helpers/configure_pr_watch.js";
import { installUserScripts } from "../helpers/install_user_scripts.js";
import { configureOmarchyMonitors } from "../helpers/configure_omarchy_monitors.js";

// URLs
const RUSTUP_URL = "https://sh.rustup.rs";
const PARU_AUR_URL = "https://aur.archlinux.org/paru.git";
const UV_INSTALL_URL = "https://astral.sh/uv/install.sh";
const FOUNDRY_INSTALL_URL = "https://foundry.paradigm.xyz";
const UOSC_INSTALL_URL =
	"https://raw.githubusercontent.com/tomasklaen/uosc/HEAD/installers/unix.sh";

// --- Constants ---
const HOME = homedir();
const _CARGO_HOME = path.join(HOME, ".cargo");
const PARU_BUILD_DIR = "/tmp/paru";
const FASTFETCH_CONFIG_DIR = path.join(HOME, ".config", "fastfetch");
// Project paths (resolved from script location, works from any cwd)
const PROJECT_ROOT = path.resolve(__dirname, "..", "..");
const COMMON_DIR = path.join(PROJECT_ROOT, "common");
const CONFIGS_DIR = path.join(PROJECT_ROOT, "configs");

const PARU_APPLIST_PATH = path.join(COMMON_DIR, "paru_applist.txt");
const FLATPAK_APPLIST_PATH = path.join(COMMON_DIR, "flatpacks_arch.txt");
const CUSTOM_FASTFETCH_CONFIG_PATH = path.join(
	CONFIGS_DIR,
	"fastfetch",
	"config.jsonc",
);

// --- Helper Functions ---

export async function resolveAurHelper(commandExistsImpl = commandExists) {
	for (const helper of ["yay", "paru"]) {
		if (await commandExistsImpl(helper)) return helper;
	}
	return null;
}

export function selectArchInstallCommand(pkg, inRepository, aurHelper) {
	if (inRepository) {
		return `sudo pacman -S --needed --noconfirm ${pkg}`;
	}
	return aurHelper ? `${aurHelper} -S --needed --noconfirm ${pkg}` : null;
}

async function packageInRepository(pkg) {
	const proc = Bun.spawn(["pacman", "-Si", pkg], {
		stdout: "ignore",
		stderr: "ignore",
	});
	return (await proc.exited) === 0;
}

export async function installGamingPackages({
	aurHelper,
	isOmarchy,
	commandExistsImpl = commandExists,
	runCommandImpl = runCommand,
} = {}) {
	const repositoryOk = await runCommandImpl(
		"sudo pacman -S --needed --noconfirm steam gamemode lib32-gamemode gamescope mangohud lib32-mangohud",
	);
	const protonOk = aurHelper
		? await runCommandImpl(
				`${aurHelper} -S --needed --noconfirm protonup-rs-bin`,
			)
		: false;

	let gpuOk = true;
	if (
		isOmarchy &&
		(await commandExistsImpl("omarchy-install-gaming-gpu-lib32"))
	) {
		gpuOk = await runCommandImpl("omarchy-install-gaming-gpu-lib32");
	} else if (!isOmarchy) {
		log.info("Skipping GPU-specific 32-bit packages outside Omarchy.");
	}
	return Boolean(repositoryOk && protonOk && gpuOk);
}

async function refreshSudo() {
	log.info("Checking sudo access. You may be prompted for your password.");
	return await runCommand("sudo -v");
}

/**
 * Snapshot pacman's local DB once so callers can pre-filter "already installed"
 * packages without paying paru's per-package AUR HTTP roundtrip. AUR packages
 * installed via paru land in the pacman DB the same way repo packages do.
 * Returns an empty Set on failure so callers fall back to "try to install
 * everything", matching pre-optimization behavior.
 */
async function getInstalledPackages() {
	try {
		const proc = Bun.spawn(["pacman", "-Qq"], {
			stdout: "pipe",
			stderr: "pipe",
		});
		const output = await new Response(proc.stdout).text();
		await proc.exited;
		return new Set(
			output
				.split("\n")
				.map((s) => s.trim())
				.filter(Boolean),
		);
	} catch (_e) {
		log.warning("pacman -Qq failed; skipping pre-install filter.");
		return new Set();
	}
}

async function installPackagesFromFile(filePath, installerCmd) {
	if (!fs.existsSync(filePath)) {
		log.warning(`Package file not found at ${filePath}`);
		return;
	}

	const content = fs.readFileSync(filePath, "utf-8");
	const requested = content
		.split("\n")
		.map((line) => line.trim())
		.filter((line) => line && !line.startsWith("#"));

	if (requested.length === 0) return;

	// Pre-filter against pacman's local DB so paru isn't invoked for packages
	// that are already installed — the big saving is skipping AUR HTTP lookups
	// on re-runs. `--needed` on the install command catches the edge cases this
	// exact-name match misses (e.g. a package satisfying a `provides` we want).
	// Only applied when installerCmd routes through pacman/paru.
	let toInstall = requested;
	let skipped = 0;
	if (installerCmd.includes("paru") || installerCmd.includes("pacman")) {
		const installed = await getInstalledPackages();
		toInstall = requested.filter((pkg) => !installed.has(pkg));
		skipped = requested.length - toInstall.length;
	}

	if (skipped > 0) {
		log.info(
			`${skipped} of ${requested.length} packages already installed — skipping.`,
		);
	}

	if (toInstall.length === 0) {
		log.success("All requested packages already installed.");
		return;
	}

	log.info(`Installing ${toInstall.length} packages individually...`);
	const failed = [];

	for (const pkg of toInstall) {
		const ok = await runCommand(`${installerCmd} ${pkg}`);
		if (ok) {
			log.success(`Installed ${pkg}`);
		} else {
			log.warning(`Failed to install ${pkg}, continuing...`);
			failed.push(pkg);
		}
	}

	if (failed.length > 0) {
		log.warning(
			`${failed.length} package(s) failed to install:\n  ${failed.join("\n  ")}`,
		);
	}
}

// --- Installation Functions ---

async function installBaseDependencies() {
	log.info("Refreshing keyrings...");
	// // Initialize and populate keys first to ensure we can verify signatures
	// await runCommand("sudo pacman-key --init");
	// await runCommand("sudo pacman-key --populate archlinux cachyos");
	// // Then update the keyring packages
	// await runCommand("sudo pacman -Sy --noconfirm archlinux-keyring cachyos-keyring");

	log.info(
		"Installing base-devel and git (required for makepkg / paru build)...",
	);
	const baseDevelOk = await runCommand(
		"sudo pacman -S --needed --noconfirm base-devel git",
	);
	if (!baseDevelOk) {
		log.warning(
			"base-devel/git install failed — paru build and AUR installs may fail.",
		);
	}

	log.info("Installing Rust via rustup...");
	await withSpinner("Installing Rust", () =>
		runCommand(`curl ${RUSTUP_URL} -sSf | sh -s -- -y`),
	);
}

async function installAurHelper() {
	if (await commandExists("paru")) {
		log.info("Paru is already installed.");
		return;
	}
	log.info("Installing paru...");
	// Remove any leftover build dir from a prior failed run so git clone succeeds.
	fs.rmSync(PARU_BUILD_DIR, { recursive: true, force: true });
	const paruOk = await withSpinner("Installing paru", () =>
		runCommand(
			`git clone ${PARU_AUR_URL} ${PARU_BUILD_DIR} && cd ${PARU_BUILD_DIR} && makepkg -si --noconfirm`,
		),
	);
	if (!paruOk) {
		log.warning(
			"paru installation failed — AUR package installs will be unavailable.",
		);
	}
}

async function ensureAurHelper() {
	const existing = await resolveAurHelper();
	if (existing) {
		log.info(`${existing} is available for AUR packages.`);
		return existing;
	}
	await installAurHelper();
	return await resolveAurHelper();
}

async function installDevTools() {
	if (!(await commandExists("uv"))) {
		log.info("Installing uv...");
		await withSpinner("Installing uv", () =>
			runCommand(`curl -LsSf ${UV_INSTALL_URL} | sh`),
		);
	} else {
		log.info("uv already installed.");
	}

	if (!(await commandExists("foundryup"))) {
		log.info("Installing Foundry...");
		await withSpinner("Installing Foundry", () =>
			runCommand(`curl -L ${FOUNDRY_INSTALL_URL} | bash`),
		);
	} else {
		log.info("Foundry (foundryup) already installed.");
	}
}

async function setupFlatpakRemotes() {
	if (await commandExists("flatpak")) {
		log.info("Setting up Flatpak remotes...");
		await runCommand(
			"flatpak remote-add --user --if-not-exists flathub https://dl.flathub.org/repo/flathub.flatpakrepo",
		);
		await runCommand("flatpak update --user -y");
	} else {
		log.warning("Flatpak not found. Skipping remote setup.");
	}
}

// --- Configuration Functions ---

async function enableServices() {
	if (await promptUser("Enable Bluetooth?", false)) {
		log.info("Enabling Bluetooth service...");
		await runCommand("sudo systemctl enable --now bluetooth");
	}

	if (
		(await commandExists("docker")) &&
		(await promptUser("Enable Docker?", true))
	) {
		log.info("Enabling and starting Docker service...");
		await runCommand("sudo systemctl enable --now docker");
		const user = process.env.USER;
		if (user) {
			await runCommand(`sudo usermod -aG docker ${user}`);
			log.warning(
				`User ${user} added to docker group. Please log out and back in.`,
			);
		}
	}
}

async function configureFastfetch() {
	if (!(await commandExists("fastfetch"))) {
		log.warning("Fastfetch command not found. Skipping configuration.");
		return;
	}

	log.info("Configuring Fastfetch...");
	fs.mkdirSync(FASTFETCH_CONFIG_DIR, { recursive: true });
	if (fs.existsSync(CUSTOM_FASTFETCH_CONFIG_PATH)) {
		safeCopyFile(
			CUSTOM_FASTFETCH_CONFIG_PATH,
			path.join(FASTFETCH_CONFIG_DIR, "config.jsonc"),
		);
		log.info("Fastfetch user config updated.");
	} else {
		log.warning(
			`Custom Fastfetch config not found at ${CUSTOM_FASTFETCH_CONFIG_PATH}`,
		);
	}
}

async function installSystemPackages(aurHelper, isOmarchy) {
	log.info("Preparing for package installation...");
	if (!(await refreshSudo())) {
		log.error(
			"Sudo authentication failed. Skipping sudo-dependent installations.",
		);
		return;
	}

	log.info("Installing packages from file lists...");
	const content = fs.readFileSync(PARU_APPLIST_PATH, "utf8");
	const requested = content
		.split("\n")
		.map((line) => line.trim())
		.filter((line) => line && !line.startsWith("#"));
	const failed = [];
	for (const pkg of requested) {
		const command = selectArchInstallCommand(
			pkg,
			await packageInRepository(pkg),
			aurHelper,
		);
		if (!command || !(await runCommand(command))) failed.push(pkg);
	}
	if (failed.length > 0) {
		log.warning(
			`${failed.length} package(s) failed to install:\n  ${failed.join("\n  ")}`,
		);
	}

	log.info("Installing Nerd Fonts...");
	await runCommand(
		"sudo pacman -S --needed --noconfirm ttf-jetbrains-mono-nerd",
	);

	if (await promptUser("Enable gaming configuration?", false)) {
		await installGamingPackages({ aurHelper, isOmarchy });
	}
}

async function installFlatpakApps() {
	if (!(await commandExists("flatpak"))) {
		log.info("Installing Flatpak...");
		await runCommand("sudo pacman -S flatpak --noconfirm");
	}
	await setupFlatpakRemotes();
	await installPackagesFromFile(
		FLATPAK_APPLIST_PATH,
		"flatpak install --user -y flathub",
	);
}

async function configureUserApps() {
	if (await promptUser("Configure git?", true)) {
		const { configureGit } = await import("../helpers/configure_git.js");
		await configureGit();
	}

	await configureMimeapps();
	await installUserScripts();
	try {
		await configureAudio();
	} catch (err) {
		log.warning(
			`Audio config sync failed (${err?.message ?? err}) — continuing with remaining app setup.`,
		);
	}

	configureBash();
	await configureFastfetch();

	log.info("Installing uosc for MPV...");
	await runCommand(`curl -fsSL ${UOSC_INSTALL_URL} | bash`);

	await enableServices();
	await configureClaude();
	await configureClaudeStayAwake();
	await configurePrWatch();
	await configureCodex();
	await configureAgentOs();
}

export async function runCachyOSSetup() {
	await installBaseDependencies();
	const aurHelper = await ensureAurHelper();
	await installDevTools();

	const isOmarchy = await commandExists("omarchy");
	await installSystemPackages(aurHelper, isOmarchy);
	await installFlatpakApps();
	await configureUserApps();
	if (isOmarchy) await configureOmarchyMonitors();

	log.success("Arch setup finished. Please restart your terminal or log out.");
}
